include/hipSYCL/sycl/exception.hpp

```diff
@@ -28,111 +28,121 @@
+  std::string message(int) const override { return "hipSYCL Error"; }
+};
+} // namespace detail
 
 class context;
 
-class exception {
+enum class errc : unsigned int {
+  success = 0,
+  runtime,
+  kernel,
+  accessor,
+  nd_range,
+  event,
+  kernel_argument,
+  build,
+  invalid,
+  memory_allocation,
+  platform,
+  profiling,
+  feature_not_supported,
+  kernel_not_supported,
+  backend_mismatch
+};
+
+inline const std::error_category &sycl_category() noexcept {
+  static const detail::sycl_category _sycl_category;
+  return _sycl_category;
+}
+  
+inline std::error_code make_error_code(errc e) noexcept {
+  return {static_cast<int>(e), sycl_category()};
+}
+  
+using async_handler = std::function<void(sycl::exception_list)>;
+  
+class exception : public virtual std::exception {
 public:
-  exception(const std::string& message)
-  : _msg{message}
-  {
-    set_error_code();
-  }
+  exception() = default;
 
-  exception(const rt::result& details)
-    : _msg{details.what()}, _error_details{details}
-  {
-    set_error_code();
-  }
+  exception(std::error_code ec, const std::string& what_arg)
+    : error_code{ec}, _msg{what_arg} {}
 
-  const char *what() const
-  {
-    return _msg.c_str();
+  exception(std::error_code ec, const char* what_arg)
+    : error_code{ec}, _msg{what_arg} {}
+
+  exception(std::error_code ec)
+    : error_code{ec} {}
+
+  exception(int ev, const std::error_category& ecat,
+            const std::string& what_arg)
+    : error_code{ev, ecat}, _msg{what_arg} {}
+
+  exception(int ev, const std::error_category& ecat, const char* what_arg)
+    : error_code{ev, ecat}, _msg{what_arg} {}
+
+  exception(int ev, const std::error_category& ecat)
+    : error_code{ev, ecat} {}
+
+  // Defined in context.hpp
+  exception(context ctx, std::error_code ec, const std::string& what_arg);
+  exception(context ctx, std::error_code ec, const char* what_arg);
+  exception(context ctx, std::error_code ec);
+  exception(context ctx, int ev, const std::error_category& ecat,
+            const std::string& what_arg);
+  exception(context ctx, int ev, const std::error_category& ecat,
+            const char* what_arg);
+  exception(context ctx, int ev, const std::error_category& ecat);
+
+  const std::error_code& code() const noexcept {
+    return error_code;
+  }
+  
+  const std::error_category& category() const noexcept {
+    return error_code.category();
   }
 
-  bool has_context() const
-  {
-    return false;
+  const char* what() const noexcept override {
+    return _msg.c_str();
   }
 
-  const std::error_code& code() const noexcept
-  {
-    return error_code;
+  bool has_context() const noexcept {
+    return (_context != nullptr);
   }
 
-  // Implementation in context.hpp
+  // Defined in context.hpp
   context get_context() const;
 
 private:
-  void set_error_code(){
-    error_code = std::error_code(_error_details.info().error_code().get_code(), std::system_category());
-  }
+  std::shared_ptr<context> _context;
```

Comment:
Why do we need this `shared_ptr` here? The SYCL spec already guarantees "common reference semantics" (which is effectively shared_ptr-like behavior) for context.